File: src/monitor/Parser.ts
import { Capture } from './Capture';

export class Parser {

    public parse(message : string) : Capture {

        const totalMinutes =  Number.parseInt(message.substring(3, 5));
        const totalSeconds =  Number.parseInt(message.substring(5, 7));
        const distance = Number.parseInt(message.substring(7, 12));
        const to500mMinutes = Number.parseInt(message.substring(13, 15));
        const to500mSeconds = Number.parseInt(message.substring(15, 17));
        const strokesPerMinute = Number.parseInt(message.substring(17, 20));
        const watt = Number.parseInt(message.substring(20, 23));
        const caloriesPerHour = Number.parseInt(message.substring(23, 27));
        const level = Number.parseInt(message.substring(27, 29));

        return {
            elapsedTime: totalMinutes*60 + totalSeconds,
            time500m: to500mMinutes*60 + to500mSeconds,
            distance: distance,
            level: level,
            watts: watt,
            cals: caloriesPerHour,
            strokesPerMinute: strokesPerMinute
        }
    }
}

File: internal/serialmonitor/parser.go
package serialmonitor

import (
	"errors"
	"math"
	"strconv"
)

func parse(line string) capture {
	totalMinutes, _ := strconv.ParseUint(line[3:5], 10, 8)
	totalSeconds, _ := strconv.ParseUint(line[5:7], 10, 8)
	elapsedTime, _ := getSecondsUint16(totalMinutes, totalSeconds)
	distance, _ := strconv.ParseUint(line[7:12], 10, 16)
	to500mMinutes, _ := strconv.ParseUint(line[13:15], 10, 8)
	to500mSeconds, _ := strconv.ParseUint(line[15:17], 10, 8)
	time500m, _ := getSecondsUint16(to500mMinutes, to500mSeconds)
	strokesPerMinute, _ := strconv.ParseUint(line[17:20], 10, 8)
	watt, _ := strconv.ParseUint(line[20:23], 10, 16)
	caloriesPerHour, _ := strconv.ParseUint(line[23:27], 10, 16)
	level, _ := strconv.ParseUint(line[27:29], 10, 8)

	return capture{
		elapsedTime:      elapsedTime,
		distance:         uint16(distance),
		time500m:         time500m,
		strokesPerMinute: uint8(strokesPerMinute),
		watts:            uint16(watt),
		cals:             uint16(caloriesPerHour),
		level:            uint8(level),
	}
}

var (
	SecondsOutOfRange = errors.New("seconds out of range")
)

func getSecondsUint16(minutes uint64, seconds uint64) (uint16, error) {

	result := minutes*60 + seconds
	if result > math.MaxUint16 {
		return 0, SecondsOutOfRange
	}

	return uint16(result), nil
}
